docs/reusing-typecheck-results.md

```diff
@@ -212,54 +200,41 @@ The phases taking the longest are marked in the right. Those are assembly import
 
 ## Implementation plan
 
-The conclusion from the above is that there is a lot of potential in caching - at the same time, implementing things to the full potential right away will require serious changes in the compiler code, which will be dangerous, hard to test and harder to review. Therefore, I propose to implement the feature in stages.
-
-### What should be done
-
-**Stage 1 - omit retypecheck for plain rebuild**
+The conclusion from the above is that there is a lot of potential in caching - at the same time, implementing things to the full potential right away will require serious changes in the compiler code, which will be dangerous, hard to test and review. Therefore, I propose to implement the feature in stages.
 
-That is, after running "fsc test.fs", rerunning `fsc test.fs` will omit type-checking step. This should work for multiple files (project) as well. Everything will be happening in `main1` - after parsing, we check the cache and deserialize the typecheck info if available, passing it to `main2` right away. If not, we do the normal type check, caching the results at the end of it, prior to passing them to the next stage. This will speed up scenarios like tinkering with plain "fsc.exe" or project rebuild when the results are incorrectly invalidated.
+**Stage 1 - force-gen and use signatures**
 
-**Stage 2 - minimize retypecheck in slightly modified projects**
+We can start with always generating signature files and using them for typecheck in cases when matching implementation files don't change. The affected phase will be `main2`.
```

Comment:
I don't think this would serve well as a "phase" in the overall goal meeting the motivation.
For me, the motivation always was "For a minimalistic change within 1 function, in a single file, in a huge project without many intra-dependencies, we should get **significantly** faster".

I don't think the proposal gets there - in the end it will be equivalent to having a project with signature files, which e.g. this repository already has and therefore this repo would not see any benefit in case of changing implementation details (i.e. a change which does not need changing a .fsi file).

Considering the amount of real-world broad testing of the signature text-generation, introducing this as a background optimization is too risky. Very likely, errors coming from mismatch between generated file and the underlying source will be fixable only be people who understand .fsi files and their different syntax in the first place, making it extremely expert-oriented.


This together with foreseen potential leads me to believe that this is not a stage (i.e. not something that can act as useful evolution step), but rather a distraction.

Any strong solution must aspire to eliminate most of the parsing and type checking time, not replace it with something roughly equivalent (duration-wise).
If I could propose, I would suggest to make the "stages" based on robustness against various kinds of changes within a project (e.g. stages of utilizing intra-file dependencies via: linear ordering `>` parsed graph `>` typechecked graph > detecting impl-only changes not affecting other files)

Reply:
It's fair that FCS itself will not benefit too much from this particular change, since it's already full of signatures.

That said:
1) Even "this repo" contains a lot of code not backed with signature files
2) We don't promote sigfiles, so there might be other big repos without sig files at all to heavily benefit from this
3) If this gets properly tested, eventually it can be a step towards getting rid of sigfiles, which will be a good thing

Help me understand your proposal here. Do you mean that force generating signatures should be:
- a separate task altogether?
- later in this impl plan (e.g. after caching opt data)?
- together with later steps right away? (e.g. with caching opt data)?

Reply:
And yes, if persistence of type-checked data works correctly and cache invalidation will be eventually taught to navigate parsed dependency graphs, I do believe this might be a good motivation to stop using .fsi files just for the sake of compilation speed benefits (as it is now).

It still might have its place for a handful of selected files as API guard, but that is not something to be kept (nor generated) for the entire project.

Reply:
> None of it, I do not think generating text-based signatures on the background without user's intention is good at all.

Well that's what I meant by "separate task" then. No okay I am not dying to dive into elaborating signature generation logic in the compiler, I was just under the impression this was something we agreed on. It's fair that if we achieve significant perf improvements with other means, this will also minimize the need of explicit signature files altogether.

Would you then start with caching sigData + optData right away, or do you think comparing typechecked graphs and acting on that is a better first step, taking into account both the potential gains and implementation complexity?

And @vzarytovskii what do you think?

Reply:
Comparing typechecked graphs is for cache invalidation, in order to invalidate smallest set of files possible.

Persisting and rehydrating typecheck results (which is more than just sigData + optData in their today's forms, but enlarging the set of values to be persisted is the real work to be done by such a feature) is IMO the first step to make anything work. Even when invalidation would work by mere linear ordering initially, that would still be a good starting point.

Reply:
Addressed in the new plan.